lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+            audience, updated_after, updated_before
+        )
+
+        # HAPI.get_annotations() returns an iterable.
+        # Turn it into a tuple because we need to iterate over it multiple times.
+        annotations = tuple(annotations)
+
+        context = DigestContext(self._db, audience, annotations)
+
+        for h_userid in audience:
+            digest = context.instructor_digest(h_userid)
+
+            if not digest["total_annotations"]:
+                # This user has no activity.
+                continue
+
+            unified_user = context.unified_users[h_userid]
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            if not to_email:
+                # We don't have an email address for this user.
+                continue
+
+            self._mailchimp_service.send_template(
+                "instructor-email-digest",
+                self._sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
+
+
+@dataclass
+class UnifiedUser:
+    """All User's for a given h_userid, unified across all ApplicationInstance's."""
+
+    h_userid: str
+    users: Tuple[User]
+    email: Optional[str]
+    display_name: Optional[str]
+
+
+@dataclass
+class UnifiedCourse:
+    """All Course's for a given authority_provided_id, unified across all ApplicationInstance's."""
+
+    authority_provided_id: str
+    title: Optional[str]
+    instructors: Tuple[UnifiedUser]
+    learner_annotations: Tuple[dict]
+
+
+class DigestContext:
+    """A context/helper object for DigestService."""
+
+    def __init__(self, db, audience, annotations):
+        self._db = db
+        self._audience = audience
+        self._annotations = annotations
+        self._unified_users = None
+        self._unified_courses = None
+
+    def instructor_digest(self, h_userid):
+        """
+        Return a digest (dict of template variables) for the given instructor.
+
+        The digest will only include courses in which both:
+
+        1. The user is an instructor
+        2. There are annotations by learners
+        """
+        course_digests = []
+
+        # Remove duplicates from self.unified_courses.values().
+        unified_courses = []
+        for unified_course in self.unified_courses.values():
+            if unified_course not in unified_courses:
+                unified_courses.append(unified_course)
+
+        for unified_course in unified_courses:
+            num_annotations = len(unified_course.learner_annotations)
+
+            if not num_annotations:
+                # There was no activity in this course.
+                continue
+
+            if h_userid not in [user.h_userid for user in unified_course.instructors]:
+                # The user isn't an instructor in this course.
+                continue
```

Comment:
Previously this was `unified_course.is_instructor(db, unified_user)` which would do a DB query each time it was called. Now the `DigestContext.unified_courses[i].instructors` list is based on DB queries that're only done once, the first time `DigestContext.unified_users` is read.